### DISCUSSION TITLE
I want to link the protobuf library in OMNeT++, but I'm encountering some problems.

### DISCUSSION THREAD

**Discussion by Jumpay** [omnetpp/omnetpp#1003] [Q&A] [answered]
I want to **link the protobuf library in OMNeT++**, but I'm encountering some problems. 
When checking the dependent version of OMNeT++ in the shell, I found that it appears to be **gcc 12.1.0 and C++17**. However, when I create an **OMNeT++ project**, I always get an **error** when compiling the abseil library: D:/File/protobuffer/include/absl/strings/string_view.h:53:26: error: no type named 'string_view' in the namespace 'std'. The code line is "using string_view = std::string_view;". This is puzzling as I can locate string_view.h through the index, but it's indicating that there's no such type in the 'std' namespace.

Furthermore, I've also created a separate **C++ project** where "using string_view = std::string_view;" compiles without any problems.
 **I'm so curious about which version of C++ the OMNeT++ project uses, and whether it's different from that in the C++ project.**

This issue has been baffling me for quite some time now.

**In addition, I am interested in finding out if OMNeT++ is suitable for joint simulations with artificial intelligence algorithms.** I have come across an open-source project based on OMNeT++ and RL/TensorFlow, but there are not many related projects out there, which has led me to question the direction I'm taking.

**Comment by rhornig** [comment]
This is to be expected because `std::string_view` is c++17 while OMNeT++ 6.0 uses c++14 for building. You can verify that in your `Makefile.inc` (look for the CXXFLAGS variable). You can override this in the `configure.user` file (CXXFLAGS) and then reconfigure and rebuild OMNeT++. (a tip: passing `V=1` on the make command line (i.e. `make MODE=release V=1`) dumps all the compiler and linker command lines helping to troubleshoot similar problems.

**Comment by rhornig** [comment]
As for the question with AI tools integration: absolutely. In fact we did have a presentation on such a topic on the last OMNeT++ summit. https://summit.omnetpp.org/2022

Take a look at the NNWIFI [video](https://www.youtube.com/watch?v=ed1_mxNrQYs) presentation and the slides.

**Comment by MrWrong77** [comment]
@Jumpay hello,  have u solve that problem?I want to link the protobuf into veins, but have no idea how to accomplish that.Can u offer me some guides for doing that? Thanks a lot.

**Reply by rhornig** [reply, marked as answer]
The best way to add external libraries is to add the necessary CFLAGS and LDFLAGS command options in the `src/makefrags` file. See how the OSG_LIBS and ws32_32 lib is added in that file.

**Reply by MrWrong77** [reply]
thanks for your advice.I found this .
```
ifneq (,$(findstring win32,$(PLATFORM)))
  #
  # on windows we have to link with the ws2_32 (winsock2) library as it is no longer added
  # to the omnetpp system libraries by default (as of OMNeT++ 5.1)
  #
  LDFLAGS += -lws2_32
  DEFINES += -DVEINS_INET_EXPORT
  ENABLE_AUTO_IMPORT=-Wl,--enable-auto-import
  LDFLAGS := $(filter-out $(ENABLE_AUTO_IMPORT), $(LDFLAGS))

  # FIX for clang not exporting typeinfo and other thunks on Windows
  ifeq ($(CC),clang)
    DEFFILE = $O/typeinfo.def
    LDFLAGS += $(DEFFILE)

```

But it need install the lib into system.I have linked the protobuf lib compiled with source into veins project, using makemake to cusomize the lib path and include path.Thanks again for your reponse.@rhornig

**Reply by MrWrong77** [reply]
Hi, @rhornig. I encounter a problem . The src of official Veins project use make-so. when I change it to make-lib and compile , the example cannot run any more, raise an runtime error, which class name is **veins::ObstacleControl**. I change nothing else.

> Class \"%s\" not found -- perhaps its code was not linked in, or the class wasn't registered with Register_Class(), or in the case of modules and channels, with Define_Module()/Define_Channel()
